examples/caa/caa.ipynb

```diff
@@ -0,0 +1,766 @@
+    "    logprobs = [logprob - min_logprob for logprob in logprobs]\n",
+    "    # Exponentiate and normalize\n",
+    "    probs = [math.exp(logprob) for logprob in logprobs]\n",
+    "    total = sum(probs)\n",
+    "    probs = [prob / total for prob in probs]\n",
+    "    return probs\n",
+    "\n",
+    "@dataclass\n",
+    "class TokenProb:\n",
+    "    token_id: int\n",
+    "    logprob: float\n",
+    "    text: str\n",
+    "\n",
+    "@dataclass\n",
+    "class TextProbs:\n",
+    "    text: str\n",
+    "    token_probs: list[TokenProb]\n",
+    "\n",
+    "    @property\n",
+    "    def sum_logprobs(self) -> float:\n",
+    "        return sum([tp.logprob for tp in self.token_probs])\n",
+    "\n",
+    "    def __repr__(self) -> str:\n",
+    "        return f\"TextProbs({self.text}:{self.sum_logprobs:.2f})\"\n",
+    "\n",
+    "def get_text_probs(input: str, model: Model, tokenizer: Tokenizer, ) -> TextProbs:\n",
+    "    \"\"\" Get the token-wise probabilities of a given input \"\"\"\n",
+    "    inputs = tokenizer(input, return_tensors=\"pt\")\n",
+    "    outputs = model(**inputs, output_hidden_states=False, return_dict=True)\n",
+    "    logprobs = torch.log_softmax(outputs.logits, dim=-1).detach().cpu()\n",
+    "    # collect the probability of the generated token -- probability at index 0 corresponds to the token at index 1\n",
+    "    logprobs = logprobs[:, :-1, :]\n",
+    "    target_ids = inputs.input_ids[:, 1:]\n",
+    "    # Get the probability of the subsequent token\n",
+    "    gen_logprobs = torch.gather(logprobs, 2, target_ids[:, :, None]).squeeze(-1)[0]\n",
+    "\n",
+    "    text_logprobs: list[TokenProb] = []\n",
+    "    for token, p in zip(target_ids[0], gen_logprobs):\n",
+    "        if token not in tokenizer.all_special_ids:\n",
+    "            text_logprobs.append(\n",
+    "                TokenProb(\n",
+    "                    token_id=token.item(),\n",
+    "                    text=tokenizer.decode(token),\n",
+    "                    logprob=p.item(),\n",
+    "                )\n",
+    "            )\n",
+    "    return TextProbs(text=input, token_probs=text_logprobs)\n",
+    "    \n",
+    "\n",
+    "def evaluate_model(\n",
+    "    model: Model, \n",
+    "    tokenizer: Tokenizer, \n",
+    "    dataset: Iterable[SteeringVectorTrainingSample],\n",
+    "    show_progress: bool = False\n",
+    "):\n",
+    "    \"\"\" Evaluate model on dataset and return normalized probability of correct answer \"\"\"\n",
+    "    total_pos_prob = 0.0\n",
+    "    for sample in tqdm(dataset, disable=not show_progress, desc=\"Evaluating\"):\n",
+    "        pos: TextProbs = get_text_probs(sample.positive_prompt, model, tokenizer)\n",
+    "        neg: TextProbs = get_text_probs(sample.negative_prompt, model, tokenizer)\n",
+    "        # NOTE: Technically, we should only evaluate log-probs of the answer tokens. \n",
```

Comment:
nit: I would word this differently, since it makes it sound like we're doing something that's incorrect, when we're not. If anything, CAA's stuff is more dubious that what we're doing here since this approach would work with any length of output, not just a single token difference like CAA. Maybe we can just note that this is equivalent to what CAA is doing without saying that we "should only evaluate log-probs of the answer". Even just saying that since the prompts are identical, this is equivalent should be fine IMO.

Reply:
I've edited the wording, how's it look now?